dask/dataframe/io/parquet/core.py

```diff
@@ -756,19 +757,34 @@ def to_parquet(
     path = fs._strip_protocol(path)
 
     if overwrite:
-        if _is_local_fs(fs):
-            working_dir = fs.expand_path(".")[0]
-            if path.rstrip("/") == working_dir.rstrip("/"):
-                raise ValueError(
-                    "Cannot clear the contents of the current working directory!"
-                )
         if append:
             raise ValueError("Cannot use both `overwrite=True` and `append=True`!")
+
         if fs.exists(path) and fs.isdir(path):
-            # Only remove path contents if
-            # (1) The path exists
-            # (2) The path is a directory
-            # (3) The path is not the current working directory
+            # Check for any previous parquet layers reading from a file in the
+            # output directory, since deleting those files now would result in
+            # errors or incorrect results.
+            for layer_name, layer in df.dask.layers.items():
+                if layer_name.startswith("read-parquet-") and isinstance(
+                    layer, DataFrameIOLayer
+                ):
+                    path_with_slash = path.rstrip("/") + "/"  # ensure trailing slash
```

Comment:
This seems fine, though a bit fragile to me, since layer names and input structures are mostly implementation details.

I wonder if it would make sense as an alternative to turn things around a bit:
1. If `overwrite=True`, write the results to a temporary obfuscated directory (or key-prefix)
2. Create a task to delete the old data and move the new data to its place
3. Make that task the final one in the `dsk` (we already have a checkpoint task in this operation which is either for the metadata or a placeholder)

I also thought about doing the overwriting before `to_parquet`, but that would force an undesirable synchronization point in what might be some otherwise embarrassingly parallel operations.

Reply:
Fair enough. I'd put this in the bucket of "it should be easier to introspect high-level layers", which would be nice to solve in whatever evolution/redesign ends up happening.